### ISSUE TITLE
[bug] SiWe as safe app does not call the authenticate function 

### ISSUE THREAD

**Issue by malteish** [rainbow-me/rainbowkit#1119]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### RainbowKit Version

0.12.4

### wagmi Version

0.12.4

### What works: Connecting (formerly gnosis) safe with WalletConnect
On connect modal select WalletConnect -> Official Modal -> copy link -> paste into app.safe.global WalletConnect app -> connection established -> sign login message -> approve in app.safe.global -> custom logic in pages/api/auth/[...nextauth].ts verifies the multisig has approved this message hash on-chain

### What doesn't work: The same workflow as safe app 
Rainbowkit support usage as (custom) safe app since version 0.12.4, so this is still brand new. We are excited about it, because we had been working on this very functionality for our dapp!
But when I try the workflow above as safe app, I get stuck with an open modal claiming to be "Verifying signature ...", without ever calling the backend.
![grafik](https://user-images.githubusercontent.com/102788938/228063640-f98049ef-7918-49ba-964d-45e048a010b2.png)
So the non-working flow is:
 Open app.safe.global -> open custom app -> paste dapp url -> accept disclaimers etc -> connect wallet (very smooth and nice!) -> sign message -> approve in safe -> get stuck at "Verifying signature ...", backend is never called.

### Expected Behavior

Sign in workflow with safe app calls backend with 0x as signature, just as sign in workflow with WalletConnect does.

### Steps To Reproduce

Not tested: enable an app to become a safe app by adding a manifest.json and configuring cors to allow loading of the same. Try signing the login message with WalletConnect and as safe app. Both should fail (unless the backend was modified to verify EIP1271), but only WalletConnect does fail whereas safe app gets stuck without ever proceeding to check the signature in the first place.

**Comment by malteish** [comment]
Thanks for looking into this, @DanielSinclair. We are very interested in gasless signatures, too, and happy to help exploring how it can be done.
Fixing the on-chain signature should be much easier though, so I recommend doing that first. And I agree that it sounds like the return value of signMessage is not forwarded properly - possibly because it contains "0x" as signature (which is expected).

**Comment by malteish** [comment]
After some digging I have to challenge our assumption that the bug starts with `signMessageAsync`. 
As the screenshot in the issue shows, I get stuck at "Verifying signature", which is only shown [after the signature has been received](https://github.com/rainbow-me/rainbowkit/blob/dfae4f16cb7abe2d4fbcc7f1226789a2998599c5/packages/rainbowkit/src/components/SignIn/SignIn.tsx#L98). 
This suggests the problem starts in [`authAdapter.verify()`](https://github.com/rainbow-me/rainbowkit/blob/dfae4f16cb7abe2d4fbcc7f1226789a2998599c5/packages/rainbowkit/src/components/SignIn/SignIn.tsx#L101). Which is weird because I log any call to `authorize` in `[...nextauth].ts` and don't see that it is executed.

**Comment by DanielSinclair** [comment]
@malteish Yes, you're right. Was able to isolate the issue to `next-auth` session handling. ```const { status } = useSession()``` is not updating to `authenticated` in the Safe environment. Still hunting down the exact reason, but it might be related to the iframe. It looks like the RainbowKit side is working as expected, so a custom authentication adapter that doesn't rely on next-auth sessions should continue to work.

**Comment by malteish** [comment]
So we found out that CSRF and the nonce in the SiWe message were equal when using Metamask or WalletConnect, but different when using safe app. This is because next-auth enforces sameSite by default (which is a good choice!). 

After updating cookie policy for testing, signing messages and authenticating them in the backend works. Added this to `[...nextauth].ts`:
```
cookies: {
      // WARNING: This disables the default cross-site protection for cookies!
      // These options were added to explore use of our app in a safe iframe.
      // https://www.npmjs.com/package/cookies-next
      // https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-rfc6265bis-03#section-4.1.2.7
      csrfToken: {
        name: `__Host-next-auth.csrf-token`,
        options: {
          httpOnly: true,
          sameSite: 'none', //'lax',
          path: '/',
          secure: true,
        },
      },
      sessionToken: {
        name: `__Secure-next-auth.session-token`,
        options: {
          httpOnly: true,
          sameSite: 'none',
          path: '/',
          secure: true,
        },
      },
    },
```
** WARNING: THESE SETTINGS ARE A VERY BAD CHOICE FOR PRODUCTION** in my opinion. Maybe someone with deeper knowledge in this area of expertise can chime in to either proof me wrong or suggest a better configuration.

Background:
https://github.com/nextauthjs/next-auth/issues/7051
https://github.com/nextauthjs/next-auth/issues/7077
https://next-auth.js.org/configuration/options#usesecurecookies

**Comment by DanielSinclair** [comment]
@malteish Will keep this issue open to see if a better solution is presented. I'll also relay to the Safe team to see if there is guidance on how to handle sessions/cookies in the Safe browser environment.

**Comment by nfts2me** [comment]
We're also interested on this. We wanted to add support not only for Safe, but also for AA in general.

Do you know of other session manager (e.g. iron-session) that is able to properly handle sessions within an iframe?

@DanielSinclair , did the Safe team gave a proper answer? It may be something recurrent for them for the way they work with apps as iframes.

**Comment by magiziz** [comment]
@nfts2me Our team is looking at this issue. We'll report back once we have a solution 🙏

**Comment by magiziz** [comment]
@nfts2me Yes. We're working on this right now, please bear with us 👍

**Comment by aaronmgdr** [comment]
having a similar issue related to signing messages with safes and getting the signature back in the dapp.
